A90-health-service-list-method.md

```diff
@@ -0,0 +1,82 @@
+# A90: Add List Method to gRPC Health Service
+----
+* **Author(s):** @marcoshuck
+* **Approver:** @markdroth
+* **Status:** In Review  
+* **Implemented in:** TBD
+* **Last updated:** 2025-02-22  
+* **Discussion at:** https://groups.google.com/g/grpc-io/c/tEI5G9sX0zc
+
+
+## Abstract
+
+This proposal introduces a new `List` RPC method for the Health service, allowing clients to retrieve the statuses of all monitored services or a filtered list of specific services. This feature simplifies integration with status-reporting dashboards and enhances observability for microservices.
+
+## Background
+
+The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored services and their statuses.
+
+This limitation makes it challenging for clients to aggregate health information across multiple services, particularly for use cases like publishing service statuses to dashboards such as [Cachet](https://cachethq.io/) or [Statuspage](https://www.atlassian.com/software/statuspage).
+
+It's important to keep in mind that the list of health services exposed by an application can change over the lifetime of the process.
+
+Kubernetes provides a similar capability with its `/readyz?verbose` endpoint, which lists the status of all components. This proposal aims to bring analogous functionality to the Health service, providing a unified view of service health.
+
+### Related Proposals
+
+N/A
+
+## Proposal
+
+Introduce a new `List` RPC method in the Health service with the following features:
+
+- Retrieve the health statuses of all monitored services or a filtered list of specified services.
+- Ensure the implementation is idempotent and side-effect free.
+- Provide a clear schema for the request and response to facilitate integration with external tools.
+
+### Proposed API Changes
+
+```protobuf
+message HealthListRequest {
+  repeated string services = 1; // Contains the name of the services that should be obtained. If empty, must return all.
```

Comment:
This field doesn't feel very `List`y - it's basically a bulk `Check` operation when this is set.  How important is that use case, vs. calling `Check` repeatedly?

Reply:
You're right on all accounts, the same result can be achieved by either calling `Check` repeatedly or simply sending an array of services to check in this method. I'm more than happy to remove this field or keep it and end up implementing it as some sort of `filter`.

Reply:
If you don't need it, I think I'd prefer to remove it.  Or rename to `filter` if you do need it.

Reply:
We can add it later on if we find a specific use case, I was just doing some premature optimization by adding filter capabilities.